### ISSUE TITLE
Human Readable CSV Findings Export

### ISSUE THREAD

**Issue by saryx123** [nccgroup/Scout2#235]
Hi,

Although various export are available and the html/js report offers an impressive user experience, a lack of easy to use, human readable csv export of scout2 findings is perhaps the one thing that keeps it from being, hands down, the best aws audit tool.

I may be missing something, even though i've spent many hours on the issue throwing myself at the [ export wiki entry ](https://github.com/nccgroup/Scout2/wiki/HowTo:-Exporting-and-Programmatically-Access-of-Scout2-Data). 
Or perhaps it was never the intention to make scout2 easy to use for mere mortals.

Regardless, i'd love to walk out of this experience a bit wiser.
Enlighten me, would you kindly :) 
-A.G.

**Comment by jbankston** [comment]
+1 a csv or json export would be great. Thanks.

**Comment by l01cd3v** [comment]
@saryx123, the difficulty with CSV is the formatting of the output - How would you expect the output to be formatted? What are rows/columns used for? What values are in the cells?

@jbankston, the data used by the HTML report is a JavaScript object, so with very little work you can actually parse JSON and send it wherever you want. The trick is to drop the first line of the file, then you have a pure JSON object. For example, the following would output you only data related to the findings:

```
tail -n +2 scout2-report/inc-awsconfig/aws_config.js | jq .services[].findings
```

**Comment by jbankston** [comment]
@l01cd3v That works. It seems the item values are hashed is that correct? Thanks.

**Comment by l01cd3v** [comment]
@jbankston, the item values are the path at which the details of the resources may be found. So for example, for IAM users, it will look like services.iam.users.<USER_ID>, where <USER_ID> is the ID set by AWS. 

In the VPC / Compute world, it would look like  services.<SERVICE_NAME>.regions.<REGION_NAME>.vpcs.<VPC_ID>.<RESOURCE_TYPE>.<RESOURCE_ID>. The resource ID is the ID set by AWS when available (sg-XXX, vpc-XXX, etc...). If no ID is set/exposed by the AWS service, it will be the name of the resource if the value cannot contain dots/periods. Last resort would be a md5 of the resource name.

**Comment by jbankston** [comment]
That is very helpful. I appreciate your quick response.

**Comment by saryx123** [comment]
@l01cd3v my take on this is that it is rather clear, scout2 is designed to help identify security faults with aws enviornments. To do that, a scout2 should provide (along the comprehensive platform) a report that is 
- clear(simple, informatively and instinctively titled & structured), 
- human readable (English, simple csv or paragraph format)
- concise (provide only the must-have info to begin remediation)
- actionable (simple users must be able to read it, forward it via email, etc) - .csv or .pdf format

As for the output format, rows and cells, **attached is a draft format for your convenience**. MetaSploit, AWS Trusted advisor and others provide similar exports. But they're all lacking in other/similar areas, thats why i'm here. 
[Cloud Audit Desired Format.zip](https://github.com/nccgroup/Scout2/files/1611907/Cloud.Audit.Desired.Format.zip)

If I had a scout2 export in such a format a few months ago when I was working for a bank, I wouldnt have sent it over to ops for remediation and my CISO(informative cc) that very moment;
![image](https://user-images.githubusercontent.com/17128601/34676210-c5ac433c-f494-11e7-9dd6-db2984eeed78.png)

Tell me how I can help, if you believe that this aligns with Scout2 purpose and vision.

-Alex G.

**Comment by l01cd3v** [comment]
Thanks for the feedback - This does not entirely align with Scout2's purpose & vision, but some of it does. The tool was designed from a security auditor's point of view, therefore it provides a point-in-time snapshot of the account's security posture and its output is meant to be consumed by relatively AWS-savvy users (*i.e.* this was never the intent to have Scout2 generate CISO-ready -or readable- reports).

Scout2's primary goal is to provide security insights for AWS accounts or, as you put it, "help identify security faults with AWS environments". With that in mind, I believe the HTML report does its job. I believe it meets 3 of your 4 requirements:

* clear: AFAIK, people figure out quite easily how to browse the HTML report.
* human readable: in English, using a technology familiar to the vast majority of humans (HTML).
* concise: resource parameters that aren't security-relevant are stripped out, color highlighting is used to indicate the faulty setting for each finding.

While I appreciate when people learn new things after using the tool, Scout2 was not created with an educational goal so providing actionable, remediation recommendations is out of scope. Furthermore, AWS environments are complex and the actionable part can be quite tricky; some findings may be as easy as "delete the resource" or "enable this feature", while others would require complex changes in the reviewed architecture. Determining the right course of action requires investigation by someone with a good understanding of the system/resource under review.

That being said, exporting the output of the tool to CSV shouldn't be too hard assuming it's OK to change your suggested format
```
Finding #,Service,Subject,ARN,ZONE,Verdict,Finding Details,Remediation
```

into the following:

```
Finding #,Service,Subject,ARN,Level,Rationale
```

Changes are:

* Drop the zone (inconsistently stored / not always available / not necessary to identify the resource)
* Change the verdict to be the level (danger/warning) - Scout2 does not say anything about resources in a "good" state and this would go against the concise part anyway.
* Change the "finding details" to be the "rationale" why this rule exists (which is something that already exists for a number of rules).
* Remove the remediation part for the reasons listed above.

**Note** that the finding number would not be consistent across runs, just a counter that increments at each line (not sure it's worth adding).

**Comment by saryx123** [comment]
Glad for the responsive and thoughtful consideration of my suggestions. few of my thoughts;

- Agreed on the 'good' state. While very useful and important for audits, is not in line with _concise_ characteristic.
- While the html report is, per say, _Human Readable_, it is not mainstream actionable, as environments where this tool would be needed most(enterprises, governments, financial) would block the report in most security engines, for example; It would never pass email transport. What good is a report you cant share with your ops team to mitigate or with your manager to be informed?
-While I like the html report experience a lot and have figured out it's use on my own, it's as you've said, savy user friendly. Again, the people who need to see report findings most will neglect to, because the only available format requires _some_ effort to get to know.
-Can agree on questionable effectiveness of mitigation advise, per reasons described in your reply.

Bottom line, I accept that the tools is oriented towards security auditors and that they should be able to navigate its ui on their own. With that said, this project has an opportunity to do tremendous good and distinguish itself by a small ad-hoc effort of a csv report feature(triggered by runtime argument).

Whatever you decide to do with this exchange and the project, I'd love your help with the method to achieve your suggested format Finding #,Service,Subject,ARN,Level,Rationale?
    How do I get there? Scout2Listall? What syntax?

Cheers,
-Sary

**Comment by l01cd3v** [comment]
Yeah, Scout2Listall will be the way to go, but I'll have to implement some changes to make it happen.

**Comment by saryx123** [comment]
Makes sense, that would help a lot.

Tell me if I can at all help.

**Comment by cvemula1** [comment]
@l01cd3v any update on this?
I am also looking for a solution to the similar issue. I am running 

**tail aws_config.js -n +2 | jq .services[].findings > findings.json_**

The output of this command in irregular JSON formate so I could not able to find the way to scan this json and sending email notification of findings. My whole requirement is to alert/notify respective service owner of each finding this I'm trying to read from the JSON object

Any workaround for this solution?

Thanks,

**Comment by j4v** [comment]
Transferring and keeping open as there's a lot of worthwhile content here.
